### ISSUE TITLE
Support binary for 32-bit builds

### ISSUE THREAD

**Issue by mattgodbolt** [mattgodbolt/compiler-explorer#495]
`-m32` and binary mode don't work together: https://godbolt.org/g/6pRN77 

```
/opt/compiler-explorer/gcc-7.1.0/bin/../lib/gcc/x86_64-linux-gnu/7.1.0/../../../../x86_64-linux-gnu/bin/ld: cannot find -lstdc++
/opt/compiler-explorer/gcc-7.1.0/bin/../lib/gcc/x86_64-linux-gnu/7.1.0/../../../../x86_64-linux-gnu/bin/ld: cannot find -lgcc_s
/opt/compiler-explorer/gcc-7.1.0/bin/../lib/gcc/x86_64-linux-gnu/7.1.0/../../../../x86_64-linux-gnu/bin/ld: skipping incompatible /opt/compiler-explorer/gcc-7.1.0/bin/../lib/gcc/x86_64-linux-gnu/7.1.0/libgcc.a when searching for -lgcc
/opt/compiler-explorer/gcc-7.1.0/bin/../lib/gcc/x86_64-linux-gnu/7.1.0/../../../../x86_64-linux-gnu/bin/ld: cannot find -lgcc
collect2: error: ld returned 1 exit status
```

**Comment by Mandar12** [comment]
Install g++-multilib.

**Comment by mattgodbolt** [comment]
I don't install `g++` from `apt` - it's built directly from source. Do you know what setting I might be missing to have it wor? The scripts to build it are here: https://github.com/mattgodbolt/compiler-explorer-image/blob/master/gcc/build/build.sh - I explicitly disabled multilib; perhaps I need to enable it?

**Comment by Mandar12** [comment]
I checked what config g++ installed from apt uses.
--enable-multiarch --with-arch-32=i686 --with-abi=m64 --with-multilib-list=m32,m64,mx32 --enable-multilib

**Comment by mattgodbolt** [comment]
Cool thanks I'll look into changing it. First up will be the nightly build
of GCC. If that seems to work we can rebuild the other GCCs too.

**Comment by mattgodbolt** [comment]
So, the GCC I built locally with the changes to enable multilib et al worked, but the automatic build last night failed. And not in an obvious way, either (build logs attached). Hopefully it'll "fix itself" for tomorrow, but I'll try and take a look later anyway...we'll get there!

[original_msg.txt](https://github.com/mattgodbolt/compiler-explorer/files/1371910/original_msg.txt)

**Comment by mattgodbolt** [comment]
And of course as soon as I post that I find the error:

```
              ~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~
../../../../gcc-trunk-20171010/libgo/runtime/proc.c:169:4: error: #error unknown case for SETCONTEXT_CLOBBERS_TLS
 #  error unknown case for SETCONTEXT_CLOBBERS_TLS
    ^~~~~
../../../../gcc-trunk-20171010/libgo/runtime/proc.c: In function 'runtime_gogo':
../../../../gcc-trunk-20171010/libgo/runtime/proc.c:284:2: error: implicit declaration of function 'fixcontext'; did you mean 'setcontext'? [-Werror=implicit-function-declaration]
  fixcontext(ucontext_arg(&newg->context[0]));
  ^~~~~~~~~~
  setcontext
../../../../gcc-trunk-20171010/libgo/runtime/proc.c: In function 'runtime_mstart':
../../../../gcc-trunk-20171010/libgo/runtime/proc.c:485:2: error: implicit declaration of function 'initcontext'; did you mean 'setcontext'? [-Werror=implicit-function-declaration]
  initcontext();
```

**Comment by mattgodbolt** [comment]
Fixed for gcc 7.3 onwards. Anyone suffering for this on earlier compilers...let me know and I can rebuild them..I think...
